src/raftstore/store/snap.rs

```diff
@@ -618,6 +665,987 @@ mod v1 {
+                let tmp_path = dir_path.join(format!("{}{}", filename, TMP_FILE_SUFFIX));
+                let cf_file = CfFile {
+                    cf: cf,
+                    path: path,
+                    tmp_path: tmp_path,
+                    ..Default::default()
+                };
+                cf_files.push(cf_file);
+            }
+
+            let meta_filename = format!("{}{}", prefix, META_FILE_SUFFIX);
+            let meta_path = dir_path.join(&meta_filename);
+            let meta_tmp_path = dir_path.join(format!("{}{}", meta_filename, TMP_FILE_SUFFIX));
+            let meta_file = MetaFile {
+                path: meta_path,
+                tmp_path: meta_tmp_path,
+                ..Default::default()
+            };
+
+            let mut s = Snap {
+                display_path: display_path,
+                cf_files: cf_files,
+                cf_index: 0,
+                meta_file: meta_file,
+                size_track: size_track,
+            };
+            // load snapshot meta if meta_file exists
+            if file_exists(&s.meta_file.path) {
+                try!(s.load_snapshot_meta())
+            }
+            Ok(s)
+        }
+
+        pub fn new_for_building<T: Into<PathBuf>>(dir: T,
+                                                  key: &SnapKey,
+                                                  snap: &DbSnapshot,
+                                                  size_track: Arc<RwLock<u64>>)
+                                                  -> RaftStoreResult<Snap> {
+            let mut s = try!(Snap::new(dir, key, size_track, SNAPSHOT_CFS, true));
+            try!(s.init_for_building(snap));
+            Ok(s)
+        }
+
+        pub fn new_for_sending<T: Into<PathBuf>>(dir: T,
+                                                 key: &SnapKey,
+                                                 size_track: Arc<RwLock<u64>>)
+                                                 -> RaftStoreResult<Snap> {
+            let mut s = try!(Snap::new(dir, key, size_track, SNAPSHOT_CFS, true));
+
+            if !s.exists() {
+                return Err(box_err!("snapshot file {} not exist", s.path()));
+            }
+            for cf_file in &mut s.cf_files {
+                // initialize cf file size and reader
+                if cf_file.size > 0 {
+                    let file = try!(File::open(&cf_file.path));
+                    cf_file.file = Some(file);
+                }
+            }
+            Ok(s)
+        }
+
+        pub fn new_for_receiving<T: Into<PathBuf>>(dir: T,
+                                                   key: &SnapKey,
+                                                   snapshot_meta: SnapshotMeta,
+                                                   size_track: Arc<RwLock<u64>>)
+                                                   -> RaftStoreResult<Snap> {
+            let cfs = try!(check_snapshot_meta(&snapshot_meta));
```

Comment:
`set_snapshot_meta` checks whether the size and checksum is correct when snapshot file exists. `check_snapshot_meta` checks whether the received snapshot meta has valid cf name and number. They are different.